clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java

```diff
@@ -171,6 +171,13 @@ public static int sizeInBytesUpperBound(byte magic, byte[] key, byte[] value, He
             return Records.LOG_OVERHEAD + LegacyRecord.recordSize(magic, key, value);
     }
 
+    public static int sizeInBytesUpperBound(byte magic, ByteBuffer key, ByteBuffer value, Header[] headers) {
```

Comment:
Thanks for clarifying the fact that these are slices. As discussed offline let us see if we can consolidate the logic (i.e., wrap the `byte[]` objects in a `ByteBuffer` and delegate to the other method).